### PR TITLE
rewrite: `uri query` replace operation

### LINKED ISSUE #6096: Query rewrite operations
Currently, the only way we offer to rewrite the query part of the request URI is with the `rewrite` handler (when it has a `?` in the target), or with `uri replace` to perform replacements on the entire URI.

These options are limiting, because many query rewrite operations are difficult to perform with those.

For example, it's impossible (I can't think of a way) to remove a single query key from a request without knowing all the other possible query keys and re-including them all with `&key={query.key}` in the rewrite.

It's also unwieldy to replace a specific query key with another (one way is with `uri replace` but it's awkward and error-prone).

I think the solution is to add `uri query` which supports multiple different types of operations on queries, similarly to the [`header` directive](https://caddyserver.com/docs/caddyfile/directives/header) which has `+` and `-` and so on.

```
# add an additional `foo=bar` (even if there was already `foo=baz`)
uri query +foo bar

# overwrite `foo` with `foo=bar`
uri query foo bar

# delete `foo`
uri query -foo

# rename key `foo` to `bar`
uri query foo>bar

# set a default value (if the key is not already set)
uri query ?foo bar

# replace values in `foo`, changing `bar` to `baz`
uri query foo bar baz

# do multiple operations at once
uri query {
	foo bar
	bar baz
	-baz
}
```

We may or may not actually implement all these operations, but I think most of them would be very nice to have.

Should this be its own directive instead of being added to `uri`? To be discussed.

### PR DESCRIPTION
**Pull request by armadi1809**
This PR is a followup to close #6096. It implements the replace operation for query strings.

### CODE REVIEW

**Comment by armadi1809** [comment]
@francislavoie let me know if you would like to see other test cases to make sure this is working as intended.

File: modules/caddyhttp/rewrite/caddyfile.go
Comment (francislavoie) [inline modules/caddyhttp/rewrite/caddyfile.go:217]: I don't remember how we did it for header replacement, did we have `SearchRegexp` support from Caddyfile? If not, I think what we should do is allow `~` as a prefix to the search string, and if so it becomes a regexp. That's the way we did it for the `map` directive: https://caddyserver.com/docs/caddyfile/directives/map#syntax. Maybe we can apply that to the header replacements Caddyfile parsing too.

**Review by francislavoie — CHANGES_REQUESTED** [review]
Additional test cases we should have:

- Placeholders (key, search, replacement)
- Partial replacement (only replace a subset of the existing value)
- Ensure nothing is replaced if the search is not found
- Caddyfile adapt test (to ensure it produces the expected JSON), just make a Caddyfile with as many config combinations as you can think of, don't need to be functional, just coverage for the adapter

File: modules/caddyhttp/rewrite/caddyfile.go
Comment (armadi1809) [reply modules/caddyhttp/rewrite/caddyfile.go:217]: I thin the header replacement uses the `SearchRegexp` exclusively when it come to the Caddyfile support. Do we want to go with that or use the `~` prefix?

Comment (francislavoie) [reply modules/caddyhttp/rewrite/caddyfile.go:217]: Oh :thinking: I guess that's true. Maybe we should do the same here then and only do regexp. WDYT @mholt

Comment (mholt) [reply modules/caddyhttp/rewrite/caddyfile.go]: Hmm, I don't remember why we did the header Caddyfile that way; but we could do the same here I suppose. Regex is more powerful :man_shrugging:

**Review by mholt — APPROVED** [review]
Lint error is unrelated, not sure why it is showing up here, but this feature LGTM!! Beautiful. Nice work @armadi1809

**Comment by armadi1809** [comment]
Thanks! Most of the scenarios mentioned by @francislavoie above should now be covered. The only one I am not sure about is using placeholders for the search, since we are now using regexes. It looks like neither the `rewrite` operation nor the `headers` replace operation support it. What do you guys think?

**Review by mholt — APPROVED** [review]
What if we merge this in for now, and then address that in a follow-up PR? Great work, @armadi1809 !

I will wait for @francislavoie 's final approval.

**Review by francislavoie — APPROVED** [review]
Thanks!

**Comment by mholt** [comment]
Yeah this is cool. 😎

**Comment by JeDaYoshi** [comment]
Since this PR was merged, caddy no longer starts and segfaults.

```
Mar 23 04:11:23 ten caddy[60323]: {"level":"info","ts":1711167083.951647,"msg":"using config from file","file":"/etc/caddy/Caddyfile"}
Mar 23 04:11:23 ten caddy[60323]: {"level":"info","ts":1711167083.9550269,"msg":"adapted config to JSON","adapter":"caddyfile"}
Mar 23 04:11:23 ten caddy[60323]: {"level":"info","ts":1711167083.9563556,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//[::1]:2019","//127.0.0.1:2019","//localhost:2019"]}
Mar 23 04:11:23 ten caddy[60323]: {"level":"info","ts":1711167083.95658,"logger":"http.auto_https","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
Mar 23 04:11:23 ten caddy[60323]: {"level":"info","ts":1711167083.9566038,"logger":"http.auto_https","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
Mar 23 04:11:23 ten caddy[60323]: {"level":"info","ts":1711167083.9567306,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc000601180"}
Mar 23 04:11:23 ten caddy[60323]: panic: runtime error: invalid memory address or nil pointer dereference
Mar 23 04:11:23 ten caddy[60323]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x13b2282]
Mar 23 04:11:23 ten systemd[1]: caddy.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Mar 23 04:11:23 ten systemd[1]: caddy.service: Failed with result 'exit-code'.
```

```
octo.autie.net {
        encode {
                zstd
                gzip
        }

        header {
                Permissions-Policy "interest-cohort=()"
                Referrer-Policy "same-origin"
                X-Content-Type-Options "nosniff"
                X-Frame-Options "SAMEORIGIN"
                X-XSS-Protection "0"
        }

        handle_path /api/* {
                reverse_proxy 127.0.0.1:4000
        }

        reverse_proxy 127.0.0.1:3000
}
```

Should I open an issue?

**Comment by armadi1809** [comment]
@mholt @francislavoie I just submitted a PR that fixes the issue mentioned in the comment above. I when calling the `Query` field of the `Rewrite` structure, within the provision method, I did not account for the fact that `Query` could be null. Sorry about that!

**Comment by mholt** [comment]
Thank you @armadi1809 :)

**Pull request merged**
